content/blog/20200130-jepsen.md

```diff
@@ -0,0 +1,81 @@
+
+> etcd 3.4.3 lived up to its claims for key-value operations: we observed nothing but strict-serializable consistency for reads, writes, and even multi-key transactions, during process pauses, crashes, clock skew, network partitions, and membership changes. Strict serializable behavior was the default for key-value operations.
+
+> Watches appear correct, at least over single keys. So long as compaction does not destroy historical data while a watch isn't running, watches appear to deliver every update to a key in order.
+
+Since Jepsen never "passes" software, but rather reports a lack of prohibitive issues, this is a great result from them. In general, it's much better than [other distributed databases](https://jepsen.io/analyses) tested to date.
+
+We believe two factors contribute to this positive result to the rigorous Jepsen analysis:
+
+### 1. A simple core
+
+A simple solution to a problem space usually results in a robust system. As an example, etcd's Raft implementation is one of the simplest implementations in open source. It focuses on the core state machine of Raft and avoids wall-time and I/O handling in the core library. And, in its API, etcd supports multi-key transactions, but adopts a simple transactional model to keep the system easier to comprehend.
+
+### 2. Rigorous testing practice
+
+Early in the development cycle of etcd, we set a goal to [achieve 80% unit test line coverage](https://github.com/etcd-io/etcd/issues/1467) and 90% unit test line coverage over core components like Raft and the MVCC storage engine. As the project progressed, we put effort into ensuring logical path coverage for core components by [introducing fail points](https://github.com/etcd-io/gofail). To embrace the randomness of timing, OS scheduling and asynchronous network I/Os, we also developed continuous integration tests, which run 24x7, to try and produce issues on a running system in a way that unit tests simply cannot. These tests have increased the quality of etcd significantly.
+
+## Identified Issues
+
+Of course, no system undergoes Jepsen testing without them reporting some areas where it could be better.  etcd had a few:
+
+### Lock Issues
+
+One lock implementation issue was found during the test where etcd failed to check the ownership of the lock before the pending lock API call returned.
+
+More specifically, in etcd a lock acquirer is associated with a session. Only if the session is not expired is the acquirer considered to still hold the lock. When acquiring the lock, the acquirer might not get the lock immediately if the lock is held by someone else. In that case the etcd server will put the acquirer into a queue and waits for its turn before returning the lock API call. However, the session might expire during this wait time. So even if it is the acquirer's turn to get the lock, the etcd server must check the existence of the session again before returning to the API call.
```

Comment:
I find this paragraph hard to understand. Here's my suggested wording:

"In etcd, a lock acquirer is associated with a session; the acquirer holds the lock until the session expires. When first attempting to acquire the lock, it may be held by someone else. In that case, the etcd server puts the acquirer into a queue, where it must wait until other lock holders release the lock. The problem is that the acquirer's session might expire during this wait time. The consequence of this premature expiration is that the etcd server must check the existence of the session again before returning to the API call."

Reply:
lgtm